[Beta fix][POS][Crash] NPE when clicking item during PTR

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: WOOMOB-363
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR fixes crash in POS in main items list.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

To reproduce the crash:
1. When pull-to-refresh happens, loadProducts(forceRefreshProducts = true) is called
2. This immediately clears the cache: productsCache.clear()
3. Then it emits the cached products (which will be empty): emit(ProductsResult.Cached(cachedProducts))
4. Then it starts fetching new products asynchronously
5. If a user clicks a product during this window between clearing the cache and fetching new products, they'll get a null product (you can add a delay in `WooPosProductsDataSource::loadProducts` to simulate the bug)

The issue is that we're clearing the cache too early. Instead, we should:
* Keep the old products in cache until we have new ones
* Only clear and update the cache when we have successfully fetched new products

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
Ensure that:
1. The crash is not happening
2. All the products are loaded on the main list
2. PTR and pagination work

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
Verified above.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
N/A

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Fix NPE when selecting item during PTR

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/256">22.2 ❄️</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Remove unused imports

samiuelson committed:
Add test

samiuelson committed:
Clean up code

samiuelson commented:
Do not merge until base branch is `release/22.2`

kidinov commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/woopos/home/items/WooPosItemsViewModelTest.kt:
```diff
@@ -881,6 +881,53 @@ class WooPosItemsViewModelTest {
+            assertThat(awaitItem()).isInstanceOf(WooPosItemsViewState.Content::class.java)
+        }
+
+        // Trigger PTR but don't let it complete
```
np: useless comment =)

samiuelson replied:
removed

kidinov reviewed (commented):
@samiuelson 👋 As discussed already

I think it fixes the issue, but to make sure to clear and set atomically, so `setAll` method in the cache may help

Also, I think the PTR is broken here - it's hidden before the response comes from the backend. Probably due to changes in the target branch.

And we may want to fix it in the release branch!

samiuelson committed:
Make clear & addAll atomic

samiuelson committed:
Don't emit Cached result when PTR

samiuelson committed:
Add test

samiuelson commented:
Thanks for review, @kidinov!

> I think it fixes the issue, but to make sure to clear and set atomically, so `setAll` method in the cache may help

Implemented and switched to `setAll`: 470dbd5
 
> Also, I think the PTR is broken here - it's hidden before the response comes from the backend. Probably due to changes in the target branch.

Good catch 👍 It was not broken in the target branch, but in this one. The reason was that after updating the data source logic to clear cache _after_ receiving Remote result, since the cache was not empty immediately, data source was returning Cached result (non-empty), resulting in `Content` UI state. I fixed that by preventing the emission of the Cached result in case `forceRefreshProducts` is `true`: 6222696 and added a test: 6f89aaf

> And we may want to fix it in the release branch!

This PR is targeting a branch that originates from `release/22.2`, so after the [parent PR](https://github.com/woocommerce/woocommerce-android/pull/13946) is merged the base will automatically change to `release/22.2`.

Let me know what you think.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/data/WooPosProductsInMemoryCache.kt:
```diff
@@ -39,4 +39,9 @@ class WooPosProductsInMemoryCache @Inject constructor() : WooPosProductsCache {
         productsCache.clear()
     }
+
+    override suspend fun setAll(products: List<Product>) = mutex.withLock {
```
@samiuelson 

There is a deadlock, isn't? Just try to PTR - it'll never exit from `setAll`

```
package com.woocommerce.android.ui.woopos.common.data

import com.woocommerce.android.model.Product
import com.woocommerce.android.ui.woopos.common.data.WooPosProductsCache.Companion.MAX_CACHE_SIZE
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class WooPosProductsInMemoryCache @Inject constructor() : WooPosProductsCache {
    private val mutex = Mutex()

    companion object {
        private const val INITIAL_CAPACITY = 200
        private const val LOAD_FACTOR = 0.75f
    }

    private val productsCache = LinkedHashMap<Long, Product>(INITIAL_CAPACITY, LOAD_FACTOR, true)

    override suspend fun addAll(products: List<Product>) = mutex.withLock {
        addProductsToCache(products)
    }

    override suspend fun getAll(): List<Product> = mutex.withLock {
        return productsCache.values.toList()
    }

    override suspend fun getProductById(productId: Long): Product? = mutex.withLock {
        return productsCache[productId]
    }

    override suspend fun clear() = mutex.withLock {
        productsCache.clear()
    }

    override suspend fun setAll(products: List<Product>) = mutex.withLock {
        productsCache.clear()
        addProductsToCache(products)
    }

    private fun addProductsToCache(products: List<Product>) {
        products.forEach { product ->
            productsCache[product.remoteId] = product
        }

        if (productsCache.size > MAX_CACHE_SIZE) {
            val keysToRemove = productsCache.keys.take(productsCache.size - MAX_CACHE_SIZE)
            keysToRemove.forEach { productsCache.remove(it) }
        }
    }
}

```

samiuelson replied:
Sorry 🤦 Fixed here: f4011b86ee2be9f5d6de31556e7c5d0013b82e43

kidinov replied:
Nothing to be sorry about. It's an interesting case. We can never call synchronized on mutex internal methods, which makes sense, but I have never thought about that

samiuelson committed:
Fix deadlock

samiuelson committed:
Remove comment from test

kidinov reviewed (approved):
LGTM! Thanks!

samiuelson committed:
Make `addAllInternal` private

samiuelson committed:
Merge branch 'release/22.2' into woomob-363-bugcrash-during-ptr-and-item-click

samiuelson committed:
Fix test
